### PR TITLE
new batching implementation

### PR DESCRIPTION
**Pull request by g4z3r**
I made a new GL batching system. Works good, except it was not tested on legacy api, will test it later.
There is new class `OpenGLQueue`. It has a list of batch objects and one buffer with all geometry data. Queue collects draw calls from `GLDrawBuf`, converts their arguments to vertex data and save it in buffer. At the same time queue creates batch objects, which contains information about drawing operation (length of batch, offset in index buffer and texture). If the next draw call has the same type, as the previous, it uses the last batch and only increases its `length` value. At the end of frame queue flushes data: creates VBO and EBO, fills them, attaches to VAOs and draws all batches.

### CODE REVIEW

File: src/dlangui/graphics/glsupport.d [hunk 1]
```diff
@@ -699,6 +642,24 @@ private void FillColor(uint color, Color[] buf_slice) {
     }
 }
 
+private float[] convertColors(uint[] cols) {
+    float[] colors;
+    colors.assumeSafeAppend();
+    colors.length = cols.length * 4;
+    for (uint i = 0; i < cols.length; i++) {
```
Comment (drug007) [inline src/dlangui/graphics/glsupport.d]:
```
for (uint i; 0..cols.length) {
```
would be more D idiomatic

File: src/dlangui/graphics/glsupport.d [hunk 2]
```diff
@@ -1678,95 +1442,234 @@ GLenum primitiveTypeToGL(PrimitiveType type) {
...
-    }
-    /// draw buffered items
+
+/// OpenGL GUI rendering queue. It collects gui draw calls, fills a big buffer for vertex data and draws everything
+private final class OpenGLQueue {
+
+    OpenGLBatch[] batches;
+
+    /// OpenGL batch structure - to draw several triangles in single OpenGL call
+    private struct OpenGLBatch {
```
Comment (drug007) [inline src/dlangui/graphics/glsupport.d]: I'm not sure, but shouldn't it be `static`?

Comment (g4z3r) [inline src/dlangui/graphics/glsupport.d]: For what purpose?

File: src/dlangui/graphics/glsupport.d (hunk 1 above)
Comment (g4z3r) [inline src/dlangui/graphics/glsupport.d]: Changed in next commit, thank you.

File: src/dlangui/graphics/glsupport.d (hunk 2 above)
Comment (drug007) [inline src/dlangui/graphics/glsupport.d]: Nested non-static aggregates contain hidden pointer to outer class/struct and this increase size of the aggregate. I just don't check if this pointer is really used in code.

Comment (g4z3r) [inline src/dlangui/graphics/glsupport.d]: Size increases only in nested in function classes/structures, as I tested.

Comment (drug007) [inline src/dlangui/graphics/glsupport.d]: I see.

**Comment by buggins** [comment]
There is already OpenGLBatch class - via GLSupport.batch 
Doesn't it do the same?

**Comment by g4z3r** [comment]
No. Previously vertex buffers were filling many times within a frame. Now there is one big buffer for all vertex data, filled one time per frame with few GL calls. OpenGLBatch class now is OpenGLQueue class with OpenGLBatch nested struct.

**Comment by buggins** [comment]
Merged.
Thank you!

**Pull request merged**